be/src/exec/pipeline/exchange/mem_limited_chunk_queue.cpp

```diff
@@ -0,0 +1,528 @@
+        }
+        flushed_rows += chunk->num_rows();
+        flushed_bytes += chunk->memory_usage();
+    }
+    std::shared_ptr<spill::Block> spill_block;
+
+    TEST_SYNC_POINT_CALLBACK("MemLimitedChunkQueue::after_calculate_max_serialize_size", &max_serialize_size);
+    raw::RawString serialize_buffer;
+    serialize_buffer.resize(max_serialize_size);
+    uint8_t* buf = reinterpret_cast<uint8_t*>(serialize_buffer.data());
+    uint8_t* begin = buf;
+    // 2. serialize data
+    for (auto& chunk : chunks) {
+        for (const auto& column : chunk->columns()) {
+            buf = serde::ColumnArraySerde::serialize(*column, buf, false, _opts.encode_level);
+            RETURN_IF(buf == nullptr, Status::InternalError("serialize data error"));
+        }
+    }
+    size_t content_length = buf - begin;
+    // 3. acquire block
+    spill::AcquireBlockOptions options;
+    options.block_size = content_length;
+    options.direct_io = false;
+    options.query_id = _state->query_id();
+    options.fragment_instance_id = _state->fragment_instance_id();
+    options.name = "mcast_local_exchange";
+    options.plan_node_id = _opts.plan_node_id;
+    ASSIGN_OR_RETURN(spill_block, _opts.block_manager->acquire_block(options));
+
+    // 4. flush serialized data
+    std::vector<Slice> data;
+    data.emplace_back(serialize_buffer.data(), content_length);
+    {
+        SCOPED_TIMER(_flush_io_timer);
+        RETURN_IF_ERROR(spill_block->append(data));
+        RETURN_IF_ERROR(spill_block->flush());
+        COUNTER_UPDATE(_flush_io_count, 1);
+        COUNTER_UPDATE(_flush_io_bytes, content_length);
+    }
+
+    {
+        std::unique_lock l(_mutex);
+        block->block = spill_block;
+        block->in_mem = false;
+        // 5. clear flushed data in memory
+        for (auto& cell : block->cells) {
+            if (cell.flushed) {
+                cell.chunk.reset();
+                block->flush_chunks++;
+            }
+        }
+        block->flush_rows = flushed_rows;
+        block->flush_bytes = flushed_bytes;
+        _flushed_accumulated_rows = block->cells.back().accumulated_rows;
+        _flushed_accumulated_bytes = block->cells.back().accumulated_bytes;
+        _next_flush_block = block->next;
+        VLOG_ROW << fmt::format("flush block [{}], rows[{}], bytes[{}]", (void*)block, _flushed_accumulated_rows,
+                                _flushed_accumulated_bytes);
+        TEST_SYNC_POINT_CALLBACK("MemLimitedChunkQueue::after_flush_block", block);
+    }
+    return Status::OK();
+}
+
+Status MemLimitedChunkQueue::_submit_flush_task() {
+    auto flush_task = [this, guard = RESOURCE_TLS_MEMTRACER_GUARD(_state)](auto& yield_ctx) {
+        TEST_SYNC_POINT("MemLimitedChunkQueue::before_execute_flush_task");
+        RETURN_IF(!guard.scoped_begin(), (void)0);
+        DEFER_GUARD_END(guard);
+        auto defer = DeferOp([&]() {
+            _has_flush_io_task = false;
```

Comment:
resetting _has_flush_io_task seems should be a atomic operation, just like set it to be true before submitting io tasks.